contracts/blender/src/contract.rs

```diff
@@ -82,22 +94,114 @@ mod blender {
         ) -> Result<()> {
             self.acquire_deposit(token_id, value)?;
             self.verify_deposit(token_id, value, note, proof)?;
-            let leaf_idx = self
-                .notes
-                .add(note)
-                .map_err(|_| BlenderError::TooManyNotes)?;
-            self.merkle_roots.insert(self.notes.root(), &());
+
+            let leaf_idx = self.append_note(note)?;
 
             self.env().emit_event(Deposited {
                 token_id,
                 value,
-                note,
                 leaf_idx,
             });
 
             Ok(())
         }
 
+        /// Trigger withdraw action (see ADR for detailed description).
+        #[allow(clippy::too_many_arguments)]
+        #[ink(message, selector = 2)]
+        pub fn withdraw(
+            &mut self,
+            token_id: TokenId,
+            value: TokenAmount,
+            recipient: AccountId,
+            fee_for_caller: Option<TokenAmount>,
+            merkle_root: MerkleRoot,
+            nullifier: Nullifier,
+            new_note: Note,
+            proof: Vec<u8>,
+        ) -> Result<()> {
+            self.verify_fee(fee_for_caller, value)?;
+            self.verify_merkle_root(merkle_root)?;
+            self.verify_nullifier(nullifier)?;
+            self.verify_withdrawal(token_id, value, merkle_root, nullifier, new_note, proof)?;
+
+            let leaf_idx = self.append_note(new_note)?;
+            self.nullifiers.insert(nullifier, &());
+
+            self.withdraw_funds(token_id, value, fee_for_caller, recipient)?;
+
+            self.env().emit_event(Withdrawn {
```

Comment:
I suggest using the approach described [here](https://github.com/deuszx/bulletin-board-example/blob/d6f00049f6c07a113d0e46b6c66825f3f382f236/contracts/README.md#creating-and-emitting-events).

Reply:
done